api/spanner/push_run.go

```diff
@@ -49,3 +67,179 @@ func HandlePushRun(ctx context.Context, auth Authenticator, w http.ResponseWrite
+
+	var run shared.TestRun
+	err := client.Get(ctx, &datastore.Key{
+		Kind: "TestRun",
+		ID:   id,
+	}, &run)
+	if err != nil {
+		logger.Errorf("Failed to load TestRun entity with integral key %d", id)
+		return nil, err
+	}
+	run.ID = id
+	return &run, nil
+}
+
+// loadRunReport loads a metrics.TestResultsReport using the URL specified in
+// run.
+func loadRunReport(ctx context.Context, run *shared.TestRun) (*metrics.TestResultsReport, error) {
+	logger := shared.GetLogger(ctx)
+
+	if run.RawResultsURL == "" {
+		str := fmt.Sprintf("TestRun entity ID=%d has no RawResultsURL", run.ID)
+		logger.Errorf(str)
+		return nil, errors.New(str)
+	}
+
+	logger.Infof("Reading report from %s", run.RawResultsURL)
+
+	resp, err := http.Get(run.RawResultsURL)
+	if err != nil {
+		logger.Warningf("Failed to load raw results from \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		str := fmt.Sprintf("Non-OK HTTP status code of %d from \"%s\" for run ID=%d", resp.StatusCode, run.RawResultsURL, run.ID)
+		logger.Warningf(str)
+		return nil, errors.New(str)
+	}
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logger.Warningf("Failed to read contents of \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	var report metrics.TestResultsReport
+	err = json.Unmarshal(data, &report)
+	if err != nil {
+		logger.Warningf("Failed to unmarshal JSON from \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	if len(report.Results) == 0 {
+		str := fmt.Sprintf("Empty report from run ID=%d (%s)", run.ID, run.RawResultsURL)
+		logger.Warningf(str)
+		return nil, errors.New(str)
+	}
+
+	logger.Infof("Read report for run ID=%d", run.ID)
+
+	return &report, nil
+}
+
+// countReportResults counts the number of meaningfully distinct test results
+// detailed in report.
+func countReportResults(ctx context.Context, report *metrics.TestResultsReport) int64 {
+	count := int64(0)
+	for _, r := range report.Results {
+		if len(r.Subtests) == 0 {
+			count++
+		} else {
+			set := mapset.NewSet()
+			for _, s := range r.Subtests {
+				if set.Contains(s.Name) {
+					shared.GetLogger(ctx).Warningf("Found test \"%s\" contains duplicate subtest name \"%s\"", r.Test, s.Name)
+				} else {
+					set.Add(s.Name)
+				}
+			}
+			count += int64(set.Cardinality())
+		}
+	}
+	return count
+}
+
+// countSpannerResults counts the number of test results bound to the given
+// runID in Cloud Spanner.
+func countSpannerResults(ctx context.Context, client *spanner.Client, runID int64) (int64, error) {
+	params := map[string]interface{}{
+		"run_id": runID,
+	}
+	s := spanner.Statement{
+		SQL:    countStmt,
+		Params: params,
+	}
+
+	shared.GetLogger(ctx).Infof("Spanner query: \"%s\" with %v", countStmt, params)
+
+	itr := client.Single().WithTimestampBound(spanner.MaxStaleness(1*time.Minute)).Query(ctx, s)
+	defer itr.Stop()
```

Comment:
I'm surprised this is needed, given that we exhaust the iterator?

Reply:
[Docs](https://godoc.org/cloud.google.com/go/spanner#RowIterator.Stop) seems to imply that this should be invoked unconditionally.